docs/master/federation/reference-resolvers.md

```diff
@@ -0,0 +1,211 @@
+# Reference resolvers
+
+To enable the current subgraph to provide entities for other subgraphs, you need to implement reference resolvers. These
+reference resolvers act as helpers to enable cross-subgraph communication and provide data from one subgraph to another
+when needed. Read more about the reference resolvers in
+the [Apollo Federation docs](https://www.apollographql.com/docs/federation/v1/entities#reference-resolvers).
+
+Lighthouse will look for a class which name is equivalent to `__typename` in the
+namespace configured in `lighthouse.federation.entities_resolver_namespace`.
+
+When you need to retrieve information from subgraphs, the gateway automatically generates a request to the corresponding
+endpoint of the subgraph. More details about this can be found in
+section [Query._entities of the Apollo Federation docs](https://www.apollographql.com/docs/federation/building-supergraphs/subgraphs-overview#query_entities)
+.
+
+An example of such a request is shown below:
+
+```graphql
+{
+    _entities(representations: [{ __typename: "Foo", id: 1 }]) {
+        ... on Foo {
+            id
+        }
+    }
+}
+```
+
+## Single Entity Resolvers
+
+After validating the type `Foo` exists, Lighthouse will look for a resolver class in
+config `lighthouse.federation.entities_resolver_namespace` namespace. The resolver class is expected to contain a
+method `__invoke()` which takes a single argument: the array form of the representation.
+
+```php
+namespace App\GraphQL\ReferenceResolvers;
+
+final class Foo
+{
+    /**
+     * @param  array{__typename: string, id: int}  $representation
+     */
+    public function __invoke(array $representation)
+    {
+        // TODO return a value that matches type Foo
+    }
+}
+```
+
+The method should only return an object that has the same name as the entity, but the namespace of the object may be
+different.
+
+```php
+namespace App\GraphQL\ReferenceResolvers;
+
+use App\GraphQL\Entities\Foo as FooEntity;
+use App\Repositories\SomeFooRepository;
+use Illuminate\Support\Arr;
+
+final class Foo
+{
+    public function __invoke($representation): FooEntity
+    {
+        $id = Arr::get($representation, 'id');
+        $foo = SomeFooRepository::findBySomeQuery($id);
+
+        return FooEntity::fromArray($foo);
+    }
+}
+```
+
+It is also possible to return an object of the current class, but in such case, you need to either create a new object
+or return its copy. This is necessary because when returning more than one entity, all entities will be assigned the
+value of the last entity since the resolver's object for all entities of the same type is the same.
```

Comment:
I would not recommend this approach or describe its caveats. Of course modifying and returning the same object repeatedly won't work.